### PR TITLE
coffee: fix the upgrade commands

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
I was on a plane and I found the time to write a PoC for better upgrade commands, this also introduced a simplification of the code

Fixes https://github.com/coffee-tools/coffee/issues/223
Fixes https://github.com/coffee-tools/coffee/issues/214

### CODE REVIEW

File: coffee_cmd/src/main.rs [hunk 1]
```diff
@@ -67,10 +67,10 @@ async fn main() -> Result<(), CoffeeError> {
                 Ok(res) => {
                     spinner.and_then(|splinner| Some(splinner.finish()));
                     match res.status {
-                        UpgradeStatus::UpToDate => {
+                        UpgradeStatus::UpToDate(_, _) => {
                             term::info!("Remote repository `{}` is already up to date!", res.repo)
                         }
-                        UpgradeStatus::Updated => {
+                        UpgradeStatus::Updated(_, _) => {
```
Comment (vincenzopalazzo) [inline coffee_cmd/src/main.rs:73]: this is a comment

File: coffee_github/src/repository.rs [hunk 2]
```diff
@@ -253,14 +253,13 @@ impl Repository for Github {
...
-        let (commit, date) = get_repo_info!(repo);
-        self.git_head = Some(commit.clone());
-        self.last_activity = Some(date.clone());
+        let status = git_upgrade(&self.url.path_string, &self.branch).await?;
+        // FIXME: add this code inside a method of the enum, to be able to call this as `status.date()`
+        let (_, date) = match status {
+            UpgradeStatus::UpToDate(ref commit, ref date) => (commit, date.clone()),
+            UpgradeStatus::Updated(ref commit, ref date) => (commit, date.clone()),
+        };
+        self.last_activity = Some(date);
```
Comment (tareknaser) [inline coffee_github/src/repository.rs:262]: We should also update `self.git_head`

File: coffee_github/src/repository.rs [hunk 3]
```diff
@@ -253,14 +253,13 @@ impl Repository for Github {
...
-        let status = fast_forward(&self.url.path_string, &self.branch)?;
-        // update the git information
-        // (git HEAD and date of the last commit)
-        let repo = git2::Repository::open(&self.url.path_string)
-            .map_err(|err| error!("{}", err.message()))?;
-        let (commit, date) = get_repo_info!(repo);
-        self.git_head = Some(commit.clone());
-        self.last_activity = Some(date.clone());
+        let status = git_upgrade(&self.url.path_string, &self.branch).await?;
+        // FIXME: add this code inside a method of the enum, to be able to call this as `status.date()`
```
Comment (tareknaser) [inline coffee_github/src/repository.rs:257]: We can do this now. left a suggestion

File: coffee_github/src/utils.rs [hunk 4]
```diff
@@ -24,46 +25,21 @@ pub async fn clone_recursive_fix(repo: git2::Repository, url: &URL) -> Result<()
...
 
-        reference
-            .set_target(fetch_commit.id(), "Fast-Forward")
-            .map_err(|err| error!("{}", err.message()))?;
+    let cmd = format!("git pull origin/{branch}");
+    sh!(path, cmd);
 
-        repo.set_head(&refname)
-            .map_err(|err| error!("{}", err.message()))?;
+    let (upstream_commit, date) = get_repo_info!(repo);
```
Comment (tareknaser) [inline coffee_github/src/utils.rs:38]: That's way more straight forward.
Thanks

File: coffee_lib/src/types/mod.rs [hunk 5]
```diff
@@ -89,8 +89,8 @@ pub mod response {
 
     #[derive(Debug, Serialize, Deserialize)]
     pub enum UpgradeStatus {
-        UpToDate,
-        Updated,
+        UpToDate(String, String),
+        Updated(String, String),
     }
 
     #[derive(Debug, Serialize, Deserialize)]
```
Comment (tareknaser) [inline coffee_lib/src/types/mod.rs:96]: Let's add this now to make the code cleaner


```suggestion
    impl UpgradeStatus {
        pub fn date(&self) -> String {
            match self {
                UpgradeStatus::UpToDate(_, date) => date.clone(),
                UpgradeStatus::Updated(_, date) => date.clone(),
            }
        }

        pub fn commit_id(&self) -> String {
            match self {
                UpgradeStatus::UpToDate(commit_id, _) => commit_id.clone(),
                UpgradeStatus::Updated(commit_id, _) => commit_id.clone(),
            }
        }
    }
```

File: coffee_github/src/utils.rs [hunk 6]
```diff
@@ -24,46 +25,21 @@ pub async fn clone_recursive_fix(repo: git2::Repository, url: &URL) -> Result<()
...
-        let refname = format!("refs/heads/{}", branch);
-        let mut reference = repo
-            .find_reference(&refname)
-            .map_err(|err| error!("{}", err.message()))?;
+    let (local_commit, _) = get_repo_info!(repo);
 
-        reference
-            .set_target(fetch_commit.id(), "Fast-Forward")
-            .map_err(|err| error!("{}", err.message()))?;
+    let cmd = format!("git pull origin/{branch}");
```
Comment (tareknaser) [inline coffee_github/src/utils.rs:35]:
```suggestion
    let cmd = format!("git pull origin {branch}");
```

**Review by tareknaser — CHANGES_REQUESTED** [review]
Great job!
Left some comments.

Let's also remove [this function](https://github.com/coffee-tools/coffee/blob/master/coffee_lib/src/plugin.rs#L151-L154) because `coffee upgrade` should upgrade a whole repository not a single plugin as per our discussion

File: coffee_github/src/utils.rs (hunk 6 above)
Comment (vincenzopalazzo) [reply coffee_github/src/utils.rs:35]: Good catch

File: coffee_github/src/repository.rs (hunk 2 above)
Comment (vincenzopalazzo) [reply coffee_github/src/repository.rs:262]: uh why it was not updated before! good catch

File: coffee_lib/src/types/mod.rs (hunk 5 above)
Comment (vincenzopalazzo) [reply coffee_lib/src/types/mod.rs:96]: Thanks!

File: coffee_github/src/utils.rs [hunk 7]
```diff
@@ -24,46 +25,25 @@ pub async fn clone_recursive_fix(repo: git2::Repository, url: &URL) -> Result<()
...
-        let refname = format!("refs/heads/{}", branch);
-        let mut reference = repo
-            .find_reference(&refname)
-            .map_err(|err| error!("{}", err.message()))?;
+    let (local_commit, _) = get_repo_info!(repo);
 
-        reference
-            .set_target(fetch_commit.id(), "Fast-Forward")
-            .map_err(|err| error!("{}", err.message()))?;
+    let cmd = format!("git pull origin {branch}");
```
Comment (tareknaser) [inline coffee_github/src/utils.rs:39]: I have a concern here 
Maybe some developer force pushed a commit and now `git pull origin {branch}` would return an error

we can do 
```bash
git fetch origin
git reset --hard origin/{branch}
``` 
to account for this case

**Review by tareknaser — APPROVED** [review]
Looks good to me
`verbose` flag works as expected too

File: coffee_github/src/utils.rs (hunk 7 above)
Comment (vincenzopalazzo) [reply coffee_github/src/utils.rs:39]: we can do a `git pull origin {branch} --force` but Idk if your is another way to do the same things?

Comment (vincenzopalazzo) [reply coffee_github/src/utils.rs]: Good suggestion, thanks!

**Pull request merged**